Index: src/generic/Makefile.am
===================================================================
--- src/generic/Makefile.am	(revision 253)
+++ src/generic/Makefile.am	(working copy)
@@ -139,8 +139,11 @@
 
 # Need to worry about these for dependency but they shouldn't
 # be included into the include directory
-extra_headers = cfortran.h frontal.h arpack.h lapack_qz.h mumps.h
+extra_headers = frontal.h arpack.h lapack_qz.h mumps.h
 
+# Need fortran wrappers accesable from driver codes so add them to the nomal headers list (instead of "extra headers")
+headers += cfortran.h
+
 # Define name of library 
 libname = generic
 
